test/e2e/framework/initialize.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
```

Comment:
I still think that letting users of the framework choose which code they actually pull in is useful, instead of forcing them to take everything. But it's a considerable change, so I understand that this might not be feasible.

The alternative is to have a set of smaller, independent functions that are in separate packages based on their functionality. The E2E test suite author then still owns the task of writing his own "setup suite" function which just calls those helper functions.

Reply:
> optionally register SynchronizedBeforeSuite and SynchronizedAfterSuite callbacks

By callbacks, you mean the function arguments that these ginkgo functions require? So, that means this separate package should have some kind of constructor for SynchronizedBeforeSuite and SynchronizedAfterSuite?

> These optional modules then can live under test/e2e/framework/suite

If I understood correctly, this means I should break the current `SetupSuite` function into smaller functions and keep them in a separate package `test/e2e/framework/suite`, right?

> Incidentally, those modules then also should own the definition of the corresponding configuration options.

These configuration options are defined in TestContext, if I'm not mistaken. Should these configuration options be all moved to these separate modules or will it be a wrapper to TestContext?

> Each module should have an interface just like ginkgo's before/after suite callbacks. The only difference is that there can be a varying number of modules while Ginkgo only allows one.

So each module's "constructor" should receive an array of interfaces?